components/hab/src/scaffolding.rs

```diff
@@ -0,0 +1,167 @@
+    try!(ui.begin("Attempting autodiscovery "));
+    try!(ui.para(
+        "No scaffolding type was provided. Let's see if we can figure out \
+        what kind of application you're planning to package.",
+    ));
+    let current_path = Path::new(".");
+    if is_project_ruby(&current_path) {
+        let ident = PackageIdent::from_str("core/scaffolding-ruby")?;
+        try!(ui.begin("We've detected your app as Ruby"));
+        let _ = ui.status(Status::Using, &format!("scaffolding package: '{}'", ident));
+        try!(ui.para(""));
+        Ok(Some(ident))
+    } else if is_project_golang(&current_path) {
+        let ident = PackageIdent::from_str("core/scaffolding-go")?;
+        try!(ui.begin("We've detected your app as Golang"));
+        let _ = ui.status(Status::Using, &format!("scaffolding package: '{}'", ident));
+        try!(ui.para(""));
+        Ok(Some(ident))
+    } else if is_project_node(&current_path) {
+        let ident = PackageIdent::from_str("core/scaffolding-node")?;
+        try!(ui.begin("We've detected your app as Node.js"));
+        let _ = ui.status(Status::Using, &format!("scaffolding package: '{}'", ident));
+        try!(ui.para(""));
+        Ok(Some(ident))
+    } else {
+        try!(ui.warn(
+            "Unable to determine the type of app in your current directory",
+        ));
+        try!(ui.para(
+            "For now, we'll generate a plan with all of the available plan options and callbacks. \
+            For more documentation on plan options try passing --withdocs or visit \
+            https://www.habitat.sh/docs/reference/plan-syntax/",
+        ));
+        Ok(None)
+    }
+}
+
+fn is_project_ruby(path: &Path) -> bool {
+    if path.join("Gemfile").is_file() {
+        return true;
+    }
+
+    if let Some(parent) = path.parent() {
+        return is_project_ruby(&parent);
+    }
+
+    return false;;
+}
+
+fn is_project_golang(path: &Path) -> bool {
+    if path.join("main.go").is_file() || path.join("Godeps/Godeps.json").is_file() ||
+        path.join("vendor/vendor.json").is_file() || path.join("glide.yaml").is_file() ||
+        project_uses_gb(path).expect("Result<bool> not returned from .go file check")
+    {
+        return true;
+    }
+
+    if let Some(parent) = path.parent() {
+        return is_project_golang(&parent);
+    }
+
+    return false;;
+}
+
+fn is_project_node(path: &Path) -> bool {
+    if path.join("package.json").is_file() {
+        return true;
+    }
+
+    if let Some(parent) = path.parent() {
+        return is_project_node(&parent);
+    }
+
+    return false;;
+}
+
+fn project_uses_gb(dir: &Path) -> io::Result<(bool)> {
+    if dir.is_dir() {
+        for entry in fs::read_dir(dir)? {
+            let entry = entry?;
+            let path = entry.path();
+            if path.is_dir() {
+                project_uses_gb(&path)?;
+            } else if path.is_file() {
+                if path.extension().unwrap() == "go" {
```

Comment:
If a file doesn't have an extension this will panic. I would write this as

```
if let Some(ext) == path.extension() {
    if ext == "go" {
        return Ok(true)
    }
}
```